File: EN.py
from sys import stdin
from itertools import imap

index = -1
def bfs(src,vis,graph):
	vis[src]+=1

	for i in xrange(len(graph[src])):
		if not vis[graph[src][i]] and vis[graph[src][i]]!=-1:
			vis = bfs(graph[src][i],vis,graph)

	return vis

def findMax(src,graph,vis,maxi):
	global index
	#print src
	for i in xrange(len(graph[src])):
		if not vis[graph[src][i]] and vis[graph[src][i]]==maxi:
			index =graph[src][i] 
			return graph[src][i]
		
		findMax(graph[src][i],graph,vis,maxi)


def main():
	dstream = imap(int,stdin.read().split())
	for i in xrange(next(dstream)):
		n = next(dstream)
		m = next(dstream)

		graph = [[] for i in xrange(n+1)]
		inv_graph = [[] for i in xrange(n+1)]
		vis = [0]*(n+1)
		inv_vis = [0]*(n+1)

		for i in xrange(m):
			a = next(dstream)
			b = next(dstream)
			graph[a].append(b)
			inv_graph[b].append(a)


		vis = bfs(1,vis,graph)
		inv_vis = bfs(n,inv_vis,inv_graph)
		
		#print vis
		#print inv_vis
		
		for i in xrange(n):
			if vis[i] and inv_vis[i]:
				pass
			else:
				vis[i]=-1

		print vis
		#print ""

		#findMax(1,graph,vis,max(vis))
		print index

main()

File: EN.cpp
#include<bits/stdc++.h>
using namespace std;
typedef long long int ll;
ll index1;
ll maxi=-INT_MAX;
ll n;

void bfs(ll src,ll vis[],vector<vector<ll> >graph)
{
  //  cout<<"src="<<src<<"\n";
    vis[src]++;
    for(ll i=0;i<graph[src].size();i++)
    {
        if(graph[src][i]==n)
        {
    //        cout<<n<<"\n";
            vis[n]++;
            return;
        }

        //else if (not vis[graph[src][i]])
        {
      //      cout<<graph[src][i]<<"\n";
            bfs(graph[src][i],vis,graph);
        }

    }
}

void findMax(ll src,ll vis[],vector<vector<ll> >graph)
{
    //cout<<src<<" "<<vis[src]<<"\n";
    for(ll i=0;i<graph[src].size();i++)
    {
        if (vis[graph[src][i]]==maxi)
        {
            //cout<<graph[src][i]<<" "<<vis[graph[src][i]]<<"=maxiIndex\n";
            index1 = graph[src][i];
            return;
        }
        else if(vis[graph[src][i]]!=-1)
        {
            //cout<<graph[src][i]<<"=not max\n";
            findMax(graph[src][i],vis,graph);
            vis[graph[src][i]] = -1;
        }
    }
}

int main()
{
	ios::sync_with_stdio(0);
    #ifndef ONLINE_JUDGE
        freopen("input.txt", "r", stdin);
        freopen("output.txt","w",stdout);
    #endif

    ll t;
    cin>>t;

    while(t--)
    {
        maxi=-INT_MAX;

        ll m;
        cin>>n>>m;

        std::vector<std::vector<ll> >graph;
        std::vector<std::vector<ll> >inv_graph;

        ll vis[n+1],inv_vis[n+1];
        for(ll i=0;i<=n;i++)
        {
            std::vector<ll> v;
            
            graph.push_back(v);
            inv_graph.push_back(v);

            vis[i]=0;
            inv_vis[i]=0;

        }

        while(m--)
        {
            ll a,b;
            cin>>a>>b;

            graph[a].push_back(b);
            inv_graph[b].push_back(a);
        }

        bfs(1,vis,graph);
        bfs(n,inv_vis,inv_graph);

        for(ll i=0;i<=n;i++)
        {
//            cout<<i<<" "<<vis[i]<<" "<<inv_vis[i]<<"\n";
            if (not (vis[i]*inv_vis[i]))
                vis[i]=-1;

            maxi = max(maxi,vis[i]);
//            cout<<vis[i]<<"\n";
        }

        findMax(1,vis,graph);

        cout<<index1<<"\n";

    }
}